### PR TITLE
[WIP] Issue 239 ambassador as proxy for jupyterhub

### PR DESCRIPTION
**Pull request by kkasravi**
This is a fix for #239 and is still a WIP - but I thought I would make the PR available for review

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/305)

### CODE REVIEW

**Comment by jlewi** [comment]
@kkasravi Thanks for doing this. I've added WIP to the title just to make it easier to track. When its ready for review just update the title.

**Comment by kkasravi** [comment]
I signed it!

File: kubeflow/core/ambassador.libsonnet [hunk 1]
```diff
@@ -170,6 +170,10 @@
                       },
                     },
                   },
+                  {
+                    name: "AMBASSADOR_SINGLE_NAMESPACE",
```
Comment (jlewi) [inline kubeflow/core/ambassador.libsonnet]: What's this do? Does this mean Ambassador only handles services in this namespace?
If that's so then we should probably delete the service we created to create a forwarding rule to the K8s dashboard.

File: kubeflow/core/jupyterhub.libsonnet [hunk 2]
```diff
@@ -116,20 +116,30 @@ c.RemoteUserAuthenticator.header_name = 'x-goog-authenticated-user-email'",
...
+              "rewrite: /hub/",
+              "service: tf-hub-0." + namespace,
+            ]),
+        },  //annotations
       },
       spec: {
-        // We want a headless service so we set the ClusterIP to be None.
-        // This headless server is used by individual Jupyter pods to connect back to the Hub.
-        clusterIP: "None",
         ports: [
```
Comment (jlewi) [inline kubeflow/core/jupyterhub.libsonnet]: Why are you making this a non-headless service?

File: kubeflow/registry.yaml [hunk 3]
```diff
@@ -2,7 +2,7 @@ apiVersion: '0.1'
 kind: ksonnet.io/registry
 libraries:
   core:
-    version: master
+    version: issue_239_ambassador_as_proxy_for_jupyterhub
```
Comment (jlewi) [inline kubeflow/registry.yaml]: FWIW this is what I do for debugging
https://github.com/kubeflow/kubeflow/blob/master/developer_guide.md

**Comment by jlewi** [comment]
@kkasravi Could you provide an update on this PR?  Is it working? Is there anything blocking it?

**Comment by kkasravi** [comment]
Update - We're finally going to test with our own cluster and going thru the iap instructions/setup. We didn't have permissions internally to do this before. I was still having some problems with jupyterhub using ambassador's /user/<uname> proxy as an external proxy that was created dynamically. I might close this PR since a number of changes have been moved in under other PRs. I'll open a new one with fewer changes. In any case I'll update by mid-week on how this is going.

**Comment by jlewi** [comment]
Great thanks.
You could also just rebase this PR to avoid changes that have already been merged.

**Comment by kkasravi** [comment]
@jlewi I created an IAP enabled cluster and integration does disable jupyterhub's default proxy (configurablehttpproxy) and moves the proxy to envoy.  Jupyterhub will use the authentication established by IAP and bring up a notebook w/o challenging the user.  So, I would recommend that both this PR and the related issue be closed. Most of the changes in this PR have been integrated in other PRs. General identity management (for example using github oauth2 instead of IAP) is an open issue #266 which I've been working on and will pickup there.

**Comment by kkasravi** [comment]
Per my above comments (IAP does proxy requests which allow jupyterhub to use it's auth and bypass it's dummyauthenticator) I"m closing this PR.

**Pull request closed without merging**